File: submissions/aiquity-for-efficient-courts/src/causelist/access.py
"""What the people in the case can see, and what they can do about it.

A schedule is also a promise to litigants and advocates. Two plain questions:

**Visibility — does each person know what is happening to their matter?**
* date certainty: was the matter actually called on the day they were told to come?
* time window: were they given a real appointment window instead of an all-day wait?
* prerequisites known in advance: when something was missing (summons, warrant, filing), was the
  matter held back with the missing item named, or did people travel only to find out in court?
* reason shown: every listing carries a stated reason and every next date a stated rule.

**Possibility — what can they do, and what does it cost them?**
* hours of life per hearing that moved the case (travel + waiting), and hours saved vs today's practice
* wasted trips (came to court, nothing moved)
* actions open to them: confirm readiness ahead (check-in), be told of a missing prerequisite
  before travelling, get a short firm date after a time request instead of an open-ended one.

Assumptions (stated): a round trip to court costs ``TRAVEL_HOURS``; without a window a person
waits ``ALL_DAY_WAIT_HOURS``; with a window they wait on average half of it.
"""
from __future__ import annotations

from collections import defaultdict

TRAVEL_HOURS = 2.0          # round trip to court, per person per listing
ALL_DAY_WAIT_HOURS = 5.0    # waiting without a time window (called at an unknown point of the day)
PEOPLE_PER_LISTING = 2      # one person for each side attends (litigant or their representative)


def _window_hours(listing) -> float:
    return max(0.25, (listing.end_min - listing.start_min) / 60.0)


def summarise(res) -> dict:
    """Visibility and possibility numbers for one simulated run (JSON-serialisable)."""
    cfg = res.config
    windows = bool(cfg.give_appointments) and cfg.planner != "baseline"
    listed = called = with_window = 0
    prereq_on_day = prereq_foreseen = 0
    wasted_trips = trips = 0
    wait_hours = 0.0
    moved = 0
    by_party: dict[str, dict] = defaultdict(lambda: {"trips": 0, "wasted": 0, "hours": 0.0, "moved": 0})
    party_of = {c.case_id: c.party_id for c in res.initial}
    party_of.update({c.case_id: c.party_id for c in res.cases})

    for d in res.days:
        outs = {o.case_id: o for o in d.outcomes}
        prereq_foreseen += sum(1 for _, why in d.plan.held_back
                               if why.startswith("prerequisite") or why.startswith("checklist"))
        for l in d.plan.listings:
            o = outs.get(l.case_id)
            if o is None:
                continue
            listed += 1
            trips += PEOPLE_PER_LISTING
            called += o.kind != "not_reached"
            wait = (_window_hours(l) / 2) if windows else ALL_DAY_WAIT_HOURS
            with_window += windows
            wait_hours += wait * PEOPLE_PER_LISTING
            if o.kind == "not_ready":
                prereq_on_day += 1
            if o.kind == "substantive":
                moved += 1
            else:
                wasted_trips += PEOPLE_PER_LISTING
            p = by_party[party_of.get(l.case_id, l.case_id)]
            p["trips"] += 1
            p["wasted"] += o.kind != "substantive"
            p["hours"] += TRAVEL_HOURS + wait
            p["moved"] += o.kind == "substantive"

    # prerequisite problems caught before anyone travelled: counted by the simulator per attempt
    counted = sum(int(c.meta.get("prereq_foreseen", 0)) for c in res.cases)
    estimated = False
    if counted:
        prereq_foreseen = max(prereq_foreseen, counted)
    elif cfg.use_readiness and cfg.planner != "baseline" and prereq_on_day:
        vis = cfg.readiness_visibility
        prereq_foreseen = round(prereq_on_day * vis / max(1e-6, 1 - vis))
        estimated = True
    life_hours = trips * TRAVEL_HOURS + wait_hours
    prereq_total = prereq_on_day + prereq_foreseen
    parties = sorted(({"party": k, **v, "hours": round(v["hours"], 1)} for k, v in by_party.items()),
                     key=lambda r: -r["hours"])
    return {
        "visibility": {
            "date_certainty_pct": round(100 * called / max(listed, 1), 1),
            "time_window_pct": round(100 * with_window / max(listed, 1), 1),
            "prerequisites_known_in_advance_pct": round(100 * prereq_foreseen / max(prereq_total, 1), 1),
            "reason_shown_pct": 100.0 if cfg.planner != "baseline" else 0.0,
            "prerequisites_foreseen": prereq_foreseen,
            "prerequisites_failed_on_the_day": prereq_on_day,
            "prerequisites_count_estimated": estimated,
        },
        "possibility": {
            "life_hours_total": round(life_hours),
            "life_hours_per_hearing_moved": round(life_hours / max(moved, 1), 1),
            "wasted_trips": wasted_trips,
            "wasted_trip_pct": round(100 * wasted_trips / max(trips, 1), 1),
            "mean_wait_hours": round(wait_hours / max(trips, 1), 2),
            "actions_open": [a for a, on in [
                ("confirm readiness ahead (check-in)", cfg.planner != "baseline"),
                ("told of a missing prerequisite before travelling", cfg.use_readiness and cfg.planner != "baseline"),
                ("real appointment window", windows),
                ("short, firm date after a time request", cfg.planner != "baseline"),
                ("date published days ahead", bool(getattr(cfg, "use_horizon", False)) and cfg.planner != "baseline"),
            ] if on],
        },
        "assumptions": {"travel_hours": TRAVEL_HOURS, "all_day_wait_hours": ALL_DAY_WAIT_HOURS,
                        "people_per_listing": PEOPLE_PER_LISTING},
        "parties_most_burdened": parties[:25],
    }


def compare(ours: dict, baseline: dict) -> dict:
    """Headline differences vs today's practice."""
    o, b = ours["possibility"], baseline["possibility"]
    return {
        "life_hours_saved": b["life_hours_total"] - o["life_hours_total"],
        "life_hours_per_hearing_moved": [b["life_hours_per_hearing_moved"], o["life_hours_per_hearing_moved"]],
        "wasted_trips_avoided": b["wasted_trips"] - o["wasted_trips"],
        "date_certainty_pct": [baseline["visibility"]["date_certainty_pct"], ours["visibility"]["date_certainty_pct"]],
    }

File: submissions/aiquity-for-efficient-courts/web/src/lib/access.ts
// What people see, and what they can do. Mirrors src/causelist/access.py; uses run.access when the
// engine exports it, otherwise derives the same numbers from the day records with the same assumptions.
import type { Run } from "./types";
import { toMin } from "./format";

export const TRAVEL_HOURS = 2.0; // round trip to court, per person per listing
export const ALL_DAY_WAIT_HOURS = 5.0; // waiting without a time window
export const PEOPLE_PER_LISTING = 2; // one person for each side

export type PartyBurden = { party: string; trips: number; wasted: number; hours: number; moved: number };
export type Access = {
  visibility: {
    date_certainty_pct: number;
    time_window_pct: number;
    prerequisites_known_in_advance_pct: number;
    reason_shown_pct: number;
    prerequisites_foreseen: number;
    prerequisites_failed_on_the_day: number;
    prerequisites_count_estimated?: boolean;
  };
  possibility: {
    life_hours_total: number;
    life_hours_per_hearing_moved: number;
    wasted_trips: number;
    wasted_trip_pct: number;
    mean_wait_hours: number;
    actions_open: string[];
  };
  assumptions: { travel_hours: number; all_day_wait_hours: number; people_per_listing: number };
  parties_most_burdened: PartyBurden[];
  derived?: boolean;
};

export const ALL_ACTIONS = [
  "confirm readiness ahead (check-in)",
  "told of a missing prerequisite before travelling",
  "real appointment window",
  "short, firm date after a time request",
  "date published days ahead",
];

const r1 = (x: number) => Math.round(x * 10) / 10;

export function accessOf(run: Run): Access {
  const own = (run as Run & { access?: Access }).access;
  if (own?.visibility && own?.possibility) return own;
  const det = run.meta.config_detail ?? {};
  const baseline = run.meta.planner === "baseline";
  const windows = Boolean(det.give_appointments ?? true) && !baseline;
  const useReadiness = Boolean(det.use_readiness ?? true);
  const visibility = Number(det.readiness_visibility ?? 0.8);
  let listed = 0, called = 0, withWindow = 0, onDay = 0, foreseen = 0, wasted = 0, trips = 0, moved = 0, wait = 0;
  const by = new Map<string, PartyBurden>();
  for (const d of run.days) {
    foreseen += d.held_back.filter((h) => h.reason.startsWith("prerequisite") || h.reason.startsWith("checklist")).length;
    for (const l of d.listings) {
      const o = l.outcome;
      if (!o) continue;
      listed += 1;
      trips += PEOPLE_PER_LISTING;
      if (o.kind !== "not_reached") called += 1;
      const w = windows ? Math.max(0.25, (toMin(l.end) - toMin(l.start)) / 60) / 2 : ALL_DAY_WAIT_HOURS;
      if (windows) withWindow += 1;
      wait += w * PEOPLE_PER_LISTING;
      if (o.kind === "not_ready") onDay += 1;
      if (o.kind === "substantive") moved += 1;
      else wasted += PEOPLE_PER_LISTING;
      const key = run.cases?.[l.case_id]?.party ?? l.case_id;
      const p = by.get(key) ?? { party: key, trips: 0, wasted: 0, hours: 0, moved: 0 };
      p.trips += 1;
      p.wasted += o.kind !== "substantive" ? 1 : 0;
      p.hours += TRAVEL_HOURS + w;
      p.moved += o.kind === "substantive" ? 1 : 0;
      by.set(key, p);
    }
  }
  let estimated = false;
  if (!foreseen && useReadiness && !baseline && onDay) {
    foreseen = Math.round((onDay * visibility) / Math.max(1e-6, 1 - visibility));
    estimated = true;
  }
  const life = trips * TRAVEL_HOURS + wait;
  const parties = [...by.values()].map((p) => ({ ...p, hours: r1(p.hours) })).sort((a, b) => b.hours - a.hours);
  const on = [!baseline, useReadiness && !baseline, windows, !baseline, Boolean(det.use_horizon) && !baseline];
  return {
    visibility: {
      date_certainty_pct: r1((100 * called) / Math.max(listed, 1)),
      time_window_pct: r1((100 * withWindow) / Math.max(listed, 1)),
      prerequisites_known_in_advance_pct: r1((100 * foreseen) / Math.max(foreseen + onDay, 1)),
      reason_shown_pct: baseline ? 0 : 100,
      prerequisites_foreseen: foreseen,
      prerequisites_failed_on_the_day: onDay,
      prerequisites_count_estimated: estimated,
    },
    possibility: {
      life_hours_total: Math.round(life),
      life_hours_per_hearing_moved: r1(life / Math.max(moved, 1)),
      wasted_trips: wasted,
      wasted_trip_pct: r1((100 * wasted) / Math.max(trips, 1)),
      mean_wait_hours: Math.round((wait / Math.max(trips, 1)) * 100) / 100,
      actions_open: ALL_ACTIONS.filter((_, i) => on[i]),
    },
    assumptions: { travel_hours: TRAVEL_HOURS, all_day_wait_hours: ALL_DAY_WAIT_HOURS, people_per_listing: PEOPLE_PER_LISTING },
    parties_most_burdened: parties.slice(0, 25),
    derived: true,
  };
}

export function hoursSaved(ours: Access, base: Access): number {
  return base.possibility.life_hours_total - ours.possibility.life_hours_total;
}
